### ISSUE TITLE
S1854 returns false positive when variable is used in method function

### ISSUE THREAD

**Issue by trockefeller-pathway** [SonarSource/sonar-dotnet#4012]
### Description

S1854 returns a false positive when a local variable is used in a method function. Method functions allow for internal functions that can capture the context of the parent method and can be helpful for organizing code, and helpful for resolving other issues like S4457 when an async portion of a method needs to be split into a separate method.

### Repro steps

```csharp
using System;
using System.Threading.Tasks;

public class Program
{
	public static void Main()
	{
		DoSomethingAsync("Hello").Wait();
	}

	public static Task DoSomethingAsync(string foo)
	{
		if (foo == null)
			throw new ArgumentNullException(nameof(foo), $"{nameof(foo)} is null");
		
		var internalVariable = calculateInternalFrom(foo); // Marked as S1854 "Remove this useless assignment to local variable 'internvalVariable'
		
		return doSomethingAsync();
		
		async Task doSomethingAsync()
		{
			Console.WriteLine(internalVariable);
			await Task.Delay(10);
		}
	}

	private static string calculateInternalFrom(string foo)
	{
		return foo + " calculations!";
	}
}
```

### Expected behavior

I would not expect this scenario to cause any S1854 warning

### Actual behavior

An S1854 warning is marked at the line specified with the comment

### Known workarounds

make the internal function take in a parameter, and then pass in the parameter to the internal function.

```csharp
...
            var internalVariable = calculateInternalFrom(foo);; // no longer marked with S1854

            return doSomethingAsync(internalVariable);

            async Task doSomethingAsync(string passVariable)
            {
                Console.WriteLine(passVariable);
                await Task.Delay(10);
            }
...
```

### Related information

* Visual Studio version 15.9.04012.0
* SonarLint for Visual Studio 2017 version 4.30.0.26640
* Operating System Windows 10

**Comment by pavel-mikula-sonarsource** [comment]
Hi @trockefeller-pathway,

Thank you for reporting this case, I can confirm it as a False Positive. We have it already reported in #3719 that you can track. I'll close this one.
